### PR TITLE
generalize subset select component to handle spatial and spectral

### PR DESCRIPTION
**Pull request by kecnry**
### Description

This pull request extends upon #1156 to support both spatial and spectral subsets (including in the same dropdown, although we don't have any uses for that, yet).  Most of this is just a generalization of logic, with a new Mixin for `SpatialSubsetSelectMixin`.

For the aperture photometry plugin, the dropdowns now look like:

Screen Shot 2022-03-17 at 1 47 59 PM

Although not actually in use, a mixed dropdown would look like:

Screen Shot 2022-03-17 at 1 45 40 PM

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Did the CI pass? If not, are the failures related?
- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`?
- [x] Is a milestone set? Milestone is only currently required for PRs related to Imviz MVP.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

File: jdaviz/core/template_mixin.py
Comment (kecnry) [inline jdaviz/core/template_mixin.py:291]: @pllim - is this equivalent to the logic currently in aperture photometry's `_get_region_from_subset`?  Should that logic move here or should aperture photometry make use of this property instead of having its own logic?

Cleaning up this logic will likely help get the aperture photometry tests passing again.

File: jdaviz/core/template_mixin.py
Comment (kecnry) [inline jdaviz/core/template_mixin.py:261]: this is where the aperture photometry tests are currently tripping (see the comment below on accessing the subset object as well).  The aperture photometry test might need to be updated to remove the checks for invalid subset entries as this now raises an error (although we should make sure it doesn't put the plugin in a state it can't recover from after the error)

File: jdaviz/core/template_mixin.py
Comment (pllim) [reply jdaviz/core/template_mixin.py]: I am not sure yet. I need to refactor aperture photometry logic for big mosaic image, which means I will try to only access the array values around the aperture. I have not yet investigated if this method grabs the whole array or only what I want. I will know more after I am done with #1138 .

Comment (kecnry) [reply jdaviz/core/template_mixin.py]: what do you think would be the best path forward then?  Leave the logic as is, get the tests passing, and then handle any changes in 1138 (preferably moving any general logic into the new component)?  Or would you rather I put this PR on hold until you have that work done?

File: jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py
Comment (pllim) [inline jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py:69]: So, you removed the ability for user to manually enter a background value?

Comment (kecnry) [reply jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py:69]: no, that is still all there!  Populating the options for the dropdown are just handled by the shared component, but if its set to 'Manual', the old behavior _should_ still work.

File: jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py
Comment (kecnry) [reply jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py:68]: Ok, very good to know, I can move that logic into the component (at least until we support renaming subsets in which case we're going to have to rethink this...).

Comment (pllim) [reply jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py:68]: Yeah, we need to rethink a lot of things, including this...

* #1141

File: jdaviz/core/template_mixin.py
Comment (pllim) [reply jdaviz/core/template_mixin.py]: Is this PR going to the Mar 25 release? My answer would depend on that.

Comment (pllim) [reply jdaviz/core/template_mixin.py]: Or you can just apply to Cubeviz for now and we can worry about Imviz later. I think this PR is a huge plus for Cubeviz, which can have both spectral and spatial subsets, but not so for Imviz.

File: jdaviz/core/template_mixin.py
Comment (pllim) [inline jdaviz/core/template_mixin.py:291]: Okay, I remember now. I did what I did to aperture photometry because of this:

* #1096

File: jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py
Comment (kecnry) [inline jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py:128]: I'm not 100% sure this is necessary, but just following the logic it seemed that it might be in some cases.  @pllim?

Comment (pllim) [reply jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py:128]: I do not think it is necessary? I only trigger background because when image change, the background level changes. But the data subset is not used until calculation button is pressed.

Comment (kecnry) [reply jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py:128]: What about the scenario where the user changes `data_selected` _after_ `subset_selected`, since `_get_region_from_subset` appears to check for `self.data_selected` before returning and setting `self._selected_subset`?

Comment (pllim) [reply jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py:128]: Good point. Right now, it does not matter because of glue-viz/glue-astronomy#52 but when that is fixed, you are probably right. Maybe you just fixed a very subtle bug. 😅

File: jdaviz/configs/default/plugins/model_fitting/model_fitting.py
Comment (pllim) [inline jdaviz/configs/default/plugins/model_fitting/model_fitting.py:278]: Is there a follow-up issue for this?

File: jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py
Comment (pllim) [reply jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py:128]: The current logic crash if data is selected but there is no subset available. You need to modify it to not crash. To reproduce:

1. Open image in Imviz.
2. Open photometry plugin.
3. Select a Data.
4. A red snackbar pops up saying Subset something something failed.

File: jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.vue
Comment (pllim) [inline jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.vue:31]: I guess this is fine, though during manual testing, I did purposely set both to be same to see the residual stats after region subtracted off its own median. But I guess it doesn't have any real science use.

File: jdaviz/core/template_mixin.py
Comment (pllim) [inline jdaviz/core/template_mixin.py:173]:
```suggestion
        plugin
```

File: jdaviz/core/template_mixin.py
Comment (pllim) [inline jdaviz/core/template_mixin.py:184]:
```suggestion
        default_text : str or None
```

File: jdaviz/core/template_mixin.py
Comment (pllim) [inline jdaviz/core/template_mixin.py:187]:
```suggestion
        allowed_type : str or None
```

File: jdaviz/core/template_mixin.py
Comment (pllim) [inline jdaviz/core/template_mixin.py:311]: Didn't you say you found a workaround? Is this note still relevant?

File: jdaviz/configs/default/plugins/model_fitting/model_fitting.py
Comment (kecnry) [reply jdaviz/configs/default/plugins/model_fitting/model_fitting.py:278]: yes, although with different behavior in each plugin, refactoring to remove all this logic might be quite involved. [:cat:](https://jira.stsci.edu/browse/JDAT-2221)

File: jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.vue
Comment (kecnry) [reply jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.vue:31]: I just copied the same check from the line analysis background.  If you'd prefer the check not be here, I'm happy to remove it!

File: jdaviz/core/template_mixin.py
Comment (kecnry) [reply jdaviz/core/template_mixin.py]: I updated the comment to explain the situation a little better (it shouldn't cause issues for imviz so long as all viewers share the same subsets, but in theory could cause us issues in the future for dynamically created viewers).

File: jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py
Comment (kecnry) [reply jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py:128]: I don't see a crash in photometry, but rather the subset dropdown is empty and the rest of the plugin UI does not appear (I don't think this is different from the previous behavior).

I just "fixed" the case where you create and select a subset but then delete it.  This now defaults to an empty selection (and hides the rest of the plugin) when there is no "default entry", but otherwise sets the default entry ("Entire Spectrum" for a lot of the spectral subset dropdowns, or "Manual" for the background dropdown in photometry).

Comment (pllim) [reply jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py:128]: Sorry, by crash, I mean a red Snackbar on the viewer. You didn't see it?

Comment (kecnry) [reply jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py:128]: Ok... I probably just missed it, but I think my last commit will have accidentally fixed it (it now returns without extracting if the dropdown is empty, but doesn't show the rest of the plugin still).

File: jdaviz/core/template_mixin.py
Comment (pllim) [inline jdaviz/core/template_mixin.py:268]: xref glue-viz/glue-astronomy#30

File: jdaviz/core/template_mixin.py
Comment (pllim) [inline jdaviz/core/template_mixin.py:305]: Do we need a way to invalidate cache if we start to allow people to rename stuff?

File: jdaviz/core/template_mixin.py
Comment (kecnry) [reply jdaviz/core/template_mixin.py:321]: Aperture photometry still uses your logic to actually extract the object (although in the future I think it would be nice to migrate to use the logic in the new component), and the dropdown is populated by messages, not looping over layers.  If you can find a reproducible case that shows an issue, let me know (although I'd argue that we should fix `get_data_from_viewer` if it is failing to show items outside viewer limits or something like that... but we definitely want to avoid re-introducing broken behavior in this PR).

File: jdaviz/core/template_mixin.py
Comment (duytnguyendtn) [inline jdaviz/core/template_mixin.py:324]: If there isn't a match made (not sure why this would happen, but who knows if a state mismatch happens at some point), no action would be taken, right? Shouldn't we through an error/exception?

Comment (pllim) [reply jdaviz/core/template_mixin.py:324]: Looks like it just `return None` if there is no match.

Comment (pllim) [reply jdaviz/core/template_mixin.py:324]: Not sure if it is more correct to crash or return None if there is no match though, I'll defer to @kecnry .

Comment (kecnry) [reply jdaviz/core/template_mixin.py:324]: right now I think returning None from something this low-level is ok.  Any higher-level code can then check against None if it needs to instead of having to wrap in a try-except.  But realistically it should never happen since the choices are synced against the subsets and manually entering an invalid string raises an error in the `_selected_changed` method.

**Review by duytnguyendtn — APPROVED** [review]
Thanks for applying the suggestions! I'll merge when all the tests pass

File: jdaviz/core/template_mixin.py
Comment (kecnry) [inline jdaviz/core/template_mixin.py:227]: this is likely the cause of #1217 - when `_subset_to_dict` loops through the viewers by ref in imviz, only the original viewer has a reference.  The subset still exists in the original viewer, so it is able to get the label and type, but the color must not be set when the message is received.

File: jdaviz/core/template_mixin.py
Comment (kecnry) [inline jdaviz/core/template_mixin.py:320]: will also need to be updated when fixing #1217

File: jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py
Comment (kecnry) [reply jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py:69]: @pllim just to revisit this - are the arbitrary masks you mentioned loaded as subsets?  Can you give me an example so I can try to write other logic to filter them out from photometry?

Comment (pllim) [reply jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py:69]: Yes, a bunch of those are loaded in the [Imviz example notebook](https://github.com/spacetelescope/jdaviz/blob/main/notebooks/ImvizExample.ipynb); look for the cell that says `imviz.load_static_regions(my_regions)`. In the example, all the names start with `my_` for simplicity but in reality, you can name it whatever you want.

Comment (pllim) [reply jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py:69]: The unfortunate thing is in order to support arbitrary shapes, everything (even something that can be cleanly converted to something that is compatible with photometry aperture) is converted to a masked state that is incompatible. That is why I said we need to rethink life.

https://github.com/spacetelescope/jdaviz/blob/a9b151d1ca91af798fcfb83bfcf34d5a5d0cf1dd/jdaviz/configs/imviz/helper.py#L236

Comment (kecnry) [reply jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py:69]: Thanks, that's very helpful.  In exploring this scenario in #1291, it doesn't seem that the situation is quite as bad as we feared as other logic in the component will see these as non-spatial subsets (unfortunately that means it assumes they're spectral, but since this only happens in imviz, that distinction isn't so important).  It will definitely require more thought and testing though.

**Pull request merged**
